From 8ddc55cab85fccf328ea5aada494c3c641622b47 Mon Sep 17 00:00:00 2001
From: kph <kph@meraki.com>
Date: Mon, 25 Nov 2013 22:54:13 +0000
Subject: [PATCH]     busybox: Add support for broadcast TFTP

    Add support for tftp to the broadcast address. Using -B enables this
    option.

    Reviewed-by: Derrick Pallas <pallas@meraki.com>
---
 networking/tftp.c | 34 +++++++++++++++++++++++++---------
 1 file changed, 25 insertions(+), 9 deletions(-)

diff --git a/networking/tftp.c b/networking/tftp.c
index 60fdff2..4bfcf5d 100644
--- a/networking/tftp.c
+++ b/networking/tftp.c
@@ -97,6 +97,7 @@
 //usage:       "Transfer a file from/to tftp server\n"
 //usage:     "\n	-l FILE	Local FILE"
 //usage:     "\n	-r FILE	Remote FILE"
+//usage:     "\n	-B Broadcast TFTP is allowed"
 //usage:	IF_FEATURE_TFTP_GET(
 //usage:     "\n	-g	Get file"
 //usage:	)
@@ -159,8 +160,9 @@
 
 /* masks coming from getopt32 */
 enum {
-	TFTP_OPT_GET = (1 << 0),
-	TFTP_OPT_PUT = (1 << 1),
+	TFTP_OPT_BROADCAST = (1 << 0),
+	TFTP_OPT_GET = (1 << 1),
+	TFTP_OPT_PUT = (1 << 2),
 	/* pseudo option: if set, it's tftpd */
 	TFTPD_OPT = (1 << 7) * ENABLE_TFTPD,
 	TFTPD_OPT_r = (1 << 8) * ENABLE_TFTPD,
@@ -171,12 +173,12 @@ enum {
 
 #if ENABLE_FEATURE_TFTP_GET && !ENABLE_FEATURE_TFTP_PUT
 #define IF_GETPUT(...)
-#define CMD_GET(cmd) 1
+#define CMD_GET(cmd) (1 << 1)
 #define CMD_PUT(cmd) 0
 #elif !ENABLE_FEATURE_TFTP_GET && ENABLE_FEATURE_TFTP_PUT
 #define IF_GETPUT(...)
 #define CMD_GET(cmd) 0
-#define CMD_PUT(cmd) 1
+#define CMD_PUT(cmd) (1 << 1)
 #else
 #define IF_GETPUT(...) __VA_ARGS__
 #define CMD_GET(cmd) ((cmd) & TFTP_OPT_GET)
@@ -297,6 +299,7 @@ static int tftp_protocol(
 		/* NULL if tftp, !NULL if tftpd: */
 		len_and_sockaddr *our_lsa,
 		len_and_sockaddr *peer_lsa,
+		int broadcast_allowed,
 		const char *local_file
 		IF_TFTP(, const char *remote_file)
 #if !ENABLE_TFTP
@@ -335,6 +338,9 @@ static int tftp_protocol(
 	socket_fd = xsocket(peer_lsa->u.sa.sa_family, SOCK_DGRAM, 0);
 	setsockopt_reuseaddr(socket_fd);
 
+	setsockopt(socket_fd, SOL_SOCKET, SO_BROADCAST,
+		   (void *)&broadcast_allowed,
+		   sizeof(broadcast_allowed));
 	if (!ENABLE_TFTP || our_lsa) { /* tftpd */
 		/* Create a socket which is:
 		 * 1. bound to IP:port peer sent 1st datagram to,
@@ -587,14 +593,20 @@ static int tftp_protocol(
 		case 1:
 			if (!our_lsa) {
 				/* tftp (not tftpd!) receiving 1st packet */
-				our_lsa = ((void*)(ptrdiff_t)-1); /* not NULL */
+
 				len = recvfrom(socket_fd, rbuf, io_bufsize, 0,
 						&peer_lsa->u.sa, &peer_lsa->len);
 				/* Our first dgram went to port 69
 				 * but reply may come from different one.
 				 * Remember and use this new port (and IP) */
-				if (len >= 0)
+				if (len >= 4) {
+					opcode = ntohs( ((uint16_t*)rbuf)[0] );
+					if ((opcode == TFTP_ERROR) && broadcast_allowed) {
+						goto recv_again;
+					}
 					xconnect(socket_fd, &peer_lsa->u.sa, peer_lsa->len);
+					our_lsa = ((void*)(ptrdiff_t)-1); /* not NULL */
+				}
 			} else {
 				/* tftpd, or not the very first packet:
 				 * socket is connect()ed, can just read from it. */
@@ -776,6 +788,8 @@ int tftp_main(int argc UNUSED_PARAM, char **argv)
 # endif
 	int result;
 	int port;
+	int broadcast_allowed;
+	int opt;
 
 	INIT_G();
 
@@ -816,7 +830,7 @@ int tftp_main(int argc UNUSED_PARAM, char **argv)
 		}
 	}
 
-	getopt32(argv, "^"
+	opt = getopt32(argv, "^" "B"
 			IF_FEATURE_TFTP_GET("g") IF_FEATURE_TFTP_PUT("p")
 			"l:r:" IF_FEATURE_TFTP_BLOCKSIZE("b:")
 			IF_FEATURE_TFTP_HPA_COMPAT("m:")
@@ -840,6 +854,8 @@ int tftp_main(int argc UNUSED_PARAM, char **argv)
 	}
 # endif
 
+	broadcast_allowed = ((opt & TFTP_OPT_BROADCAST) != 0);
+
 	if (remote_file) {
 		if (!local_file) {
 			const char *slash = strrchr(remote_file, '/');
@@ -866,7 +882,7 @@ int tftp_main(int argc UNUSED_PARAM, char **argv)
 	G.file = remote_file;
 # endif
 	result = tftp_protocol(
-		NULL /*our_lsa*/, peer_lsa,
+		NULL /*our_lsa*/, peer_lsa, broadcast_allowed,
 		(LONE_DASH(local_file) ? NULL : local_file), remote_file
 		IF_FEATURE_TFTP_BLOCKSIZE(, 1 /* want_transfer_size */)
 		IF_FEATURE_TFTP_BLOCKSIZE(, blksize)
@@ -999,7 +1015,7 @@ int tftpd_main(int argc UNUSED_PARAM, char **argv)
 	close(STDIN_FILENO); /* close old, possibly wildcard socket */
 	/* tftp_protocol() will create new one, bound to particular local IP */
 	result = tftp_protocol(
-		our_lsa, peer_lsa,
+		our_lsa, peer_lsa, FALSE,
 		local_file
 		IF_TFTP(, NULL /*remote_file*/)
 		IF_FEATURE_TFTP_BLOCKSIZE(, want_transfer_size)
-- 
2.32.0

